### ISSUE TITLE
bug: After "long" running resolve -> back-button appears

### ISSUE THREAD

**Issue by pavax** [driftyco/ionic#3941]
**Type**: <span ionic-type>bug</span>

**Platform**: <span ionic-platform>all</span>  

<span ionic-description>I'm referring to the following Bug: https://github.com/driftyco/ionic/issues/3699 since it was closed automatically.

Here is a Plunker based on the ionic-starter-side-menu that shows the described behaviour in different ionic versions:

*ionic 1.0.0 BETA 1 (works) *
http://plnkr.co/edit/y6lE2Af6PoGxmuGb4kTu?p=preview

*ionic 1.0.0 (works not) *
http://plnkr.co/edit/orJT13V5txYSH144vtND?p=preview

<b>How to reproduce:</b>

Go to the 'Playlist' or 'Search' Menu Content
Click on Browse (this is the state that has a resolve that takes more than 1 sec)
An alert should show up and the browse-state is now loaded => The Back button is now shown all of a sudden
If you reduce the $timeout value in the resolve to 0 (or anything under ~ 300ms) everything works.... Unfortunately my server can not respond that fast all the time :)

<b>Last working version</b>

The last working version seems to be 1.0.0-beta.13.

1.0.0-beta.13 (working):
http://plnkr.co/edit/70q913JX3AhTC8vuGymj?p=preview

1.0.0-beta.14 (not-working):
http://plnkr.co/edit/G60ahN9wGbLXbQGIApWr?p=preview
</span>

<span is-issue-template></span>

**Comment by Ionitron** [comment]
Greetings @pavax!

My sensors indicate that you need to update your issue through our [custom issue form](http://ionicframework.com/submit-issue/#/?iid=3941). We are now requiring all issues to be submitted this way, to ensure that we have all of the information necessary to fix them as quickly as possible.

**[Click Here To Update Your Issue](http://ionicframework.com/submit-issue/#/?iid=3941)**

I will have no choice but to close this issue if it is not resubmitted through the form. Please fill out the rest of the form, so that I may use my friendly robot powers to assist you.

Thank you!

<span ionitron-issue-resubmit></span>

**Comment by nyze2oo9** [comment]
I have a similar problem. I'm using a factory with localstorage and http request and now I have implemented promises into the factory, but now everytime when I click on that view in the navigation the ham-nav disapears and back buttons shows up. Any solutions?

**Comment by sagivf** [comment]
+1

**Comment by smkhalsa** [comment]
+1 I'm also having this issue. The back button is showing when I use the resolve option.

**Comment by nyze2oo9** [comment]
I was able to solve the issue. Here is some code mabye this will help some of you. You just have to get rid of the resolve in the routing. Instead use the factory function in the controller. 

 .controller('NewsCtrl', function($scope, $http, NewsService, $ionicLoading) {
        $ionicLoading.show({
            template: 'loading...'
        });
        NewsService.getAllNews(true).then(function(data){
            $scope.news = data;
        }).finally(function() {
        $ionicLoading.hide();
    });

**Comment by sagivf** [comment]
@nyze2oo9 Thanks for the solution but I would still like to be able to use the resolve.
It allows me to globally use the loading event as well as only transition when the request is successful.
It also makes for a more testable and cleaner code in my opinion.

**Comment by pavax** [comment]
Looks like it is fixed in v1.0.1

http://plnkr.co/edit/lShmW88us19Oq3pUsB5E?p=preview

It might have been fixed due the following changes: https://github.com/driftyco/ionic/commit/cef6708

**Comment by smkhalsa** [comment]
I'm still having this issue in 1.1.0. Any suggestions?

**Comment by rubnvp** [comment]
:+1:  I'm having the same problem, :/

**Comment by mhartington** [comment]
@smkhalsa and @rubnvp do you guys have a codepen that replicates this?

**Comment by rubnvp** [comment]
Hi @mhartington you can check it here: http://codepen.io/rubnvp/pen/GJVzWZ
If you go to browse and then return to playlists the 'menu button' will be replace with the 'back button' (and should not)

**Comment by rubnvp** [comment]
ok, I just find my issue,  the resolve should be outside the views object, like this:

.state('app.playlists', {
      url: "/playlists",
      resolve: {
            data: function($timeout, $q){
              var dfd = $q.defer();

              $timeout(function(){dfd.resolve("Async resolved")},500);

              return dfd.promise;
            }
          }
      views: {
        'menuContent': {
          templateUrl: "playlists.html",
          controller: 'PlaylistsCtrl'          
        }
      }
    })

**Comment by smkhalsa** [comment]
Yes, looks like @rubnvp's codepen shows the issue I've been experiencing. When you click "search", then click "playlists", the back button shows as "search" instead of the hamburger menu.

**Comment by talamaska** [comment]
I'm having exactly the same issue. Resolve does break the history. I have lost like 3 hours to understand what is happening. I don't get the logic in this. Why a slow resolve breaks the history and makes the back button show up, what they have in common?

**Comment by adriantombu** [comment]
Aaaaaaannnnnd 4 months later, still that annoying bug \o/

**Comment by joelcdoyle** [comment]
Argh! Can't wait to see this fixed

**Comment by talamaska** [comment]
Can we resolve this ourself? I'm tired of waiting.

On Friday, November 20, 2015, Joel Doyle <notifications@github.com> wrote:

> Argh! Can't wait to see this fixed
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/driftyco/ionic/issues/3941#issuecomment-158501121>.
>


-- 
Best Regards,
Zlati Pehlivanov

**Comment by kabus202** [comment]
+1

**Comment by nhoctanker** [comment]
+1

**Comment by talamaska** [comment]
The problem is this line of code https://github.com/driftyco/ionic/blob/master/js/angular/directive/menuClose.js#L53

I'm trying to understand why is this hardcoded and how it can be fixed and replace with something more dynamic

**Comment by imasubmarine** [comment]
is there any way  to fix this temporarily in 1.1.1 ?

**Comment by beat-at-hartwig** [comment]
+1 `ionic 1.2.4` aarrrghs what a mess...

**Comment by talamaska** [comment]
https://github.com/driftyco/ionic/issues/4900

**Comment by beat-at-hartwig** [comment]
Originally my issue is occuring in a bit different situation. If my side-menu is open, I usually want to click on an item to navigate to another view. Every view has a `rootHistory = true`. While I am interacting very reasonable with the Interface everything is working as expected.

**But** when I am trying to click very fast on random items in that item-list, for example with 2-3 fingers at the same time before the `side-menu` closes within the 300ms delay, the navigation to a view is not very consistent/traceable. I am using the `$ionicView.loaded` and `$ionicView.enter` events to assign my data from the controller to the view, which leads into small delays like the `resolve` does. Doing so the `burger`-button disappears sometimes like mentioned above...

**Comment by jgw96** [comment]
Hello! Are you all still experiencing this issue?

**Comment by talamaska** [comment]
@jgw96 well I didn't try it recently, cause I remove all the resolves in my router, but i have opened a separate issue and shared my findings. You can try to replicate with the above mentioned codepen and there is a comment about  the last version 1.2.4, so my thinking is that it is still available.

**Comment by talamaska** [comment]
@jgw96 I tryed the codepen by replacing with the last version of ionic and it is indeed broken.

**Comment by xming13** [comment]
+1

**Comment by jgw96** [comment]
@talamaska thanks for the info! I

**Comment by icedice** [comment]
This is still a issue in 1.2.4 and it is release critical for me. Any workarounds or fixes yet?

**Comment by danbucholtz** [comment]
Hi @pavax,

The Ionic team has discussed this and decided not to make any changes here.  Ionic was designed without the resolve functionality of ui-router in mind.  It will often work just fine, but often times it does not.  We're sorry about that.

In order to support resolve data and view caching, we would have to start making some "hacks" to ui-router since the resolve data is injected on creation, but views are only created once (typically).  We just think the risk is too high.  There are also all kinds of changes that would have to occur with transitions and timings.

A work around would be to perform whatever logic (load data, etc) that would occur in resolve before beginning the state transition.

Thank you for contributing and we really appreciate you using Ionic.

Thanks,
Dan

**Comment by sagivf** [comment]
@danbucholtz thanks for addressing this in detail :+1: 

Not fixing this is totally reasonable as it is indeed super risky and not relevant for Ionic2.
But I would strongly recommend documenting this.
I was a long time user of the resolve functionality, both in `ui-router` and angular's default `router` (I find it makes a lot of scene). When I started using Ionic I naturally used it and it took me weeks to months to realise that my problems were with the resolve. Especially since It was working against the browser which adds a lot of confusion when your trying to figure things out.

In short this can be a major issue for users and should be documented so others don't go through the hell I went trough (really this was one of the most confusing things i've had...).

Thanks !
